hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java

```diff
@@ -194,47 +186,50 @@ private Result populateResult(Query query) {
                 queryResultSizeLimiter.getNodeResultLimit(query.getPartitionIdSet().size()));
     }
 
-    private void doRunOnPartitionThreads(Query query, PartitionIdSet partitionIds, Result result) {
-        try {
-            List<Future<Result>> futures = dispatchPartitionScanQueryOnOwnerMemberOnPartitionThread(
+    private CompletableFuture<Result> doRunOnPartitionThreads(Query query, PartitionIdSet partitionIds, Result result) {
+
+        List<CompletableFuture<Result>> futures = dispatchPartitionScanQueryOnOwnerMemberOnPartitionThread(
                     query, partitionIds);
-            addResultsOfPredicate(futures, result, partitionIds, true);
-        } catch (Throwable t) {
-            throw rethrow(t);
-        }
+        return addResultsOfPredicate(futures, result, partitionIds, true);
+
     }
 
     @SuppressWarnings("unchecked")
     // modifies partitionIds list! Optimization not to allocate an extra collection with collected partitionIds
-    private void addResultsOfPredicate(List<Future<Result>> futures, Result result,
-                                       PartitionIdSet unfinishedPartitionIds, boolean rethrowAll) {
-        for (Future<Result> future : futures) {
-            Result queryResult = null;
-
-            try {
-                queryResult = future.get();
-            } catch (Throwable t) {
-                if (t.getCause() instanceof QueryResultSizeExceededException || rethrowAll) {
-                    throw rethrow(t);
-                }
-                logger.fine("Could not get query results", t);
-            }
-
-            if (queryResult == null) {
-                continue;
-            }
-            PartitionIdSet queriedPartitionIds = queryResult.getPartitionIds();
-            if (queriedPartitionIds != null) {
-                if (!unfinishedPartitionIds.containsAll(queriedPartitionIds)) {
-                    // do not take into account results that contain partition IDs already removed from partitionIds
-                    // collection as this means that we will count results from a single partition twice
-                    // see also https://github.com/hazelcast/hazelcast/issues/6471
-                    continue;
-                }
-                unfinishedPartitionIds.removeAll(queriedPartitionIds);
-                result.combine(queryResult);
-            }
-        }
+    private CompletableFuture<Result> addResultsOfPredicate(List<CompletableFuture<Result>> futures, Result result,
+                                                            PartitionIdSet unfinishedPartitionIds, boolean rethrowAll) {
+        CompletableFuture<Result>[] futureArray = futures.toArray(new CompletableFuture[0]);
+
+        return CompletableFuture.allOf(futureArray)
+                .thenApplyAsync(v -> {
```

Comment:
i might be wrong , please correct me 
i believe if any future completes exceptionally then the thenApplyAsync wont execute as matter of fact any other Async processing on the future wont be executed and it will be propagated to the caller thread and when we join() the future it will throw Exception wrapped in CompletionException 
which is handled by peeling of the completion Exception as the underlying Exception will be actual cause.

Reply:
We were iterating futures one by one and if any of them fails we were rethrowing their error. This is clear, right? 

Let's see what we are doing in your new code: 

First of all, we are now taking the first future that is failing into account among all futures regardless of the order of them. 

Second, allOf() call will be rejected with the error of the first failing future **without** rethrowing. Plus, there is no if check, if a future is rejected we will return a future that is rejected no matter what. However, we should have checked `if (t.getCause() instanceof QueryResultSizeExceededException || rethrowAll) {`.

Therefore, IMO, I think we should remove allOf and do this instead:

 1. Using a for loop, iterate over all futures
 2. If one of the futures needs to throw (use the same if statement inside catch block), return a immediately rejected completable future 
 3. If all futures are processed and none of them needs to throw, return a successful future.

Reply:
completable future allOf() completes even if any of its future throws exception or fails exceptionally  
It means that the thenApplyAsync call will be applied to the combined future, regardless of whether individual futures completed exceptionally or not.

- have made the changes for join() to get() as requested.

Reply:
thenApplySync will run only if allOf() completableFuture completes normally. If a individual future completes exceptionally, allOf will do as well. see https://www.educative.io/answers/what-is-completablefutureallof-in-java and https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#thenApplyAsync-java.util.function.Function-